cmd-ref: plots: add templates docs (#3697)
iterative/dvc.org: content/docs/command-reference/plots/templates.md

to JSON files for you use as base for new templates. To modify them, use
any valid elements of the [Vega-Lite specification].

You can dump a specific built-in template by providing it's name as
argument, for example `dvc plots templates confusion`. Template
files are written to `.dvc/plots` by default, but any location can be
set with the `--out` (`-o`) option.

<admon type="note">

Note that templates can only be used with [data-series plots].

</admon>

[plot templates]:
  https://dvc.org/doc/command-reference/plots#plot-templates-data-series-only
[Vega-Lite specification]: https://vega.github.io/vega-lite/
[data-series plots]: /doc/command-reference/plots#types-of-metrics
  
## Options

- `-o <path>`, `--out <path>` - Directory to save templates to.

## Example: Modifying the `simple` template

The built-in `simple` template can be an ideal base for custom templates
because it has a minimal structure you can make quick modifications to. For
example, let's show vertical bars instead of a connected line.

We'll work with the following `data.csv` file:

```csv
x,y
0,0.1
1,0.4
2,0.9
3,1.6
```

The simple template renders it like this:

```dvc

> $ dvc plots show data.csv -x x -y y --template simple

Reviewer: Do we need `-x x -y y`? Seems redundant.

Author: Not necessarily, I can modify it. Do we want just a `--template` showcase?

Reviewer: > Not necessarily, I can modify it. Do we want just a --template showcase?

Yes, removing the `-x x -y y` would look better IMO. Thanks @pared

Author: Should be fixed now

Reviewer: Thanks @pared !

file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_templates_show_unmodified.svg)

Let's dump the `simple` template to the current working directory and
rename it appropriately:

```dvc
$ dvc plots templates simple -o .
Templates have been written into '.'.

$ mv simple.json bars_template.json
```

Now, let's modify the `bars_template.json` file to display the bars (instead of a
line):

```cli
$ cat bars_template.json
```

> ````git
>  {
>    "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
>    ...
>    "mark": {
> -    "type": "line"
> +    "type": "bar"
>    },
>    "encoding": { ...
>    ...
>  }```

Reviewer: Oops...
````suggestion
```git
 {
   "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
   ...
   "mark": {
-    "type": "line"
+    "type": "bar"
   },
   "encoding": { ...
   ...
 }
```
````

# plot templates

> Dump plot templates.
> 
> Sometimes users need to customize the way project data is visualized. This
> command allows dumping templates used by DVC, so that the users don't have to
> build their own templates from zero.

Reviewer: ```suggestion
Dump built-in plot templates to JSON files you can customize.

```

## Synopsis

```usage
usage: dvc plots templates [-h] [-q | -v] [-o <path>]
       [{simple,linear,confusion,confusion_normalized,scatter,smooth}]

positional arguments:

>  TEMPLATE
>                         Template to write. Writes all templates by default.

Reviewer: ```suggestion
 TEMPLATE    Template to write. Writes all templates by default.
```

```

## Description

> Users have the ability to change the way data-series plots are displayed by
> modifying the [Vega-Lite specification](https://vega.github.io/vega-lite/), thus
> generating plots in the style that best fits their needs. This keeps <abbr>DVC
> projects</abbr> programming language agnostic, as it's independent from user
> display configuration and visualization code.

Reviewer: ```suggestion
Some times you may need to customize the way `dvc plots` are rendered beyond
what the built-in [plot templates] allow.`dvc plots templates` can dump these templates
to JSON files for you use as base for new templates. To modify them, use
any valid elements of the [Vega-Lite specification].
```

> To make creation of custom plots easier, DVC provides the `templates` command
> that can dump one or all templates used by DVC.
> 
> `templates` command dumps the templates to `.dvc/plots` by default.

Reviewer: ```suggestion
You can dump a specific built-in template by providing it's name as
argument, for example `dvc plots templates confusion`. Template
files are written to `.dvc/plots` by default, but any location can be
set with the `--out` (`-o`) option.
```

> Note that templates can only be used with
> [data-series plots](/doc/command-reference/plots#description).

Reviewer: ```suggestion
<admon type="note">

Note that templates can only be used with
[data-series plots](/doc/command-reference/plots#types-of-metrics).

</admon>
```

> 
> ## Options

Reviewer: ```suggestion

[plot templates]:
  https://dvc.org/doc/command-reference/plots#plot-templates-data-series-only
[Vega-Lite specification]: https://vega.github.io/vega-lite/
  
## Options
```

## Synopsis

```usage
usage: dvc plots templates [-h] [-q | -v] [-o <path>]
       [{simple,linear,confusion,confusion_normalized,scatter,smooth}]

positional arguments:
 TEMPLATE    Template to write. Writes all templates by default.
```

## Description

Some times you may need to customize the way `dvc plots` are rendered beyond
what the built-in [plot templates] allow.`dvc plots templates` can dump these templates
to JSON files for you use as base for new templates. To modify them, use
any valid elements of the [Vega-Lite specification].

You can dump a specific built-in template by providing it's name as
argument, for example `dvc plots templates confusion`. Template
files are written to `.dvc/plots` by default, but any location can be
set with the `--out` (`-o`) option.

<admon type="note">

Note that templates can only be used with
[data-series plots](/doc/command-reference/plots#types-of-metrics).

</admon>

[plot templates]:
  https://dvc.org/doc/command-reference/plots#plot-templates-data-series-only
[Vega-Lite specification]: https://vega.github.io/vega-lite/
  
## Options

- `-o <path>`, `--out <path>` - Directory to save templates to.

> ## Example: Modifying simple template
> 
> We will modify the `simple` template to show bars instead of line. Lets take
> sample `data.csv` file:
> 
> ```
> x,y

Reviewer: ```suggestion
## Example: Modifying the `simple` template

The built-in `simple` template can be an ideal base for custom templates
because it has a minimal structure you can make quick modifications to. For
example, let's show vertical bars instead of a connected line.

We'll work with the following `data.csv` file:

```csv
x,y
```

0,0.1
1,0.4
2,0.9
3,1.6
```

> And try to visualize it:

Reviewer: ```suggestion
The simple template renders it like this:
```

```dvc
$ dvc plots show data.csv -x x -y y --template simple
file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_templates_show_unmodified.svg)

> Let's dump the simple template and rename it:
> 
> ```dvc
> $ dvc plots templates simple -o .
> Templates have been written into '.'.
> 
> $ mv simple.json custom_template.json

Reviewer: ````suggestion
Let's dump the `simple` template to the current working directory and
rename it appropriately:

```dvc
$ dvc plots templates simple -o .
Templates have been written into '.'.

$ mv simple.json bars_template.json
````

```

> Now, lets modify the `custom_template.json` so that we will use bar plot instead
> of linear:
> 
> ```cli
> $ cat custom_template.json

Reviewer: ```suggestion
Now, let's modify the `bars_template.json` file to display the bars (instead of a
line):

```cli
$ cat bars_template.json
```

```

> ```json
> {
>   "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
>   "data": {
>     "values": "<DVC_METRIC_DATA>"
>   },
>   "title": "<DVC_METRIC_TITLE>",
>   "width": 300,
>   "height": 300,
>   "mark": {
>     "type": "bar" // <<<  there used to be "line" here
>   },
>   "encoding": {
>     "x": {
>       "field": "<DVC_METRIC_X>",
>       "type": "quantitative",
>       "title": "<DVC_METRIC_X_LABEL>"
>     },
>     "y": {
>       "field": "<DVC_METRIC_Y>",
>       "type": "quantitative",
>       "title": "<DVC_METRIC_Y_LABEL>",
>       "scale": {
>         "zero": false
>       }
>     },
>     "color": {
>       "field": "rev",
>       "type": "nominal"
>     }
>   }
> }
> ```

Reviewer: Let's see how this looks in the review app... ⏳ 
```suggestion
````git
 {
   "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
   ...
   "mark": {
-    "type": "line"
+    "type": "bar"
   },
   "encoding": { ...
   ...
 }```
````

Reviewer: Sample template update looks good!

![image](https://user-images.githubusercontent.com/1477535/177469507-3f644ff1-6fe1-4b42-82f6-bb7911630a7a.png)

Although... I'm not sure it should be the `+` line first or after... ⏳

Reviewer: After is correct. Resolving

> Create visualization with custom template:
> 
> ```dvc
> $ dvc plots show data.csv -x x -y y --template custom_template.json

Reviewer: ```suggestion
And this is how the data looks like using our custom template:

```dvc
$ dvc plots show data.csv -x x -y y --template bars_template.json
```

> ## plot templates

Reviewer 2: ```suggestion
# plot templates
```

Dump plot templates.

Sometimes users need to customize the way project data is visualized. This
command allows dumping templates used by DVC, so that the users don't have to
build their own templates from zero.

## Synopsis

```usage

> usage: dvc plots templates [-h] [-q | -v] [-o <path>] [{simple,linear,confusion,confusion_normalized,scatter,smooth}]

Reviewer 2: ```suggestion
usage: dvc plots templates [-h] [-q | -v] [-o <path>]
                           [{simple,linear,confusion,confusion_normalized,scatter,smooth}]
```

> usage: dvc plots templates [-h] [-q | -v] [-o <path>] [{simple,linear,confusion,confusion_normalized,scatter,smooth}]

Reviewer 2: Copying from https://github.com/iterative/dvc.org/pull/3691#discussion_r907907003:

> yes, it's more about this command's interface in this case. Feels like this list `{simple,linear,confusion,confusion_normalized,scatter,smooth}` will be growing / changing. Why do we have to specify it on the parser level, it's not very common to limit that large number of targets there.

Reviewer 2: Opened https://github.com/iterative/dvc/issues/7944

positional arguments:

>  template

Reviewer: I see this was [suggested into lower case](https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1020932156) but to be consistent with the `COMMAND` text we use in other refs I think it should be upper case here too:

```suggestion
 TEMPLATE
```

                        Template to write. Writes all templates by default.
```

## Description

Users have the ability to change the way data-series plots are displayed by
modifying the [Vega-Lite specification](https://vega.github.io/vega-lite/), thus
generating plots in the style that best fits their needs. This keeps <abbr>DVC
projects</abbr> programming language agnostic, as it's independent from user
display configuration and visualization code.

> To make creation of custom plots easier, DVC provides `templates` command that

Reviewer 2: ```suggestion
To make creation of custom plots easier, DVC provides the `templates` command that
```

can dump one or all templates used by DVC.

`templates` command dumps the templates to `.dvc/plots` by default.

Note that templates can only be used with
[data-series plots](/doc/command-reference/plots#description).

### DVC template anchors

Plot templates are [Vega-Lite](https://vega.github.io/vega-lite/) JSON
specifications. They use predefined DVC anchors as placeholders for DVC to
inject the plot values.

- `<DVC_METRIC_DATA>` (**required**) - the plot data from any type of metrics
  files is converted to a single JSON array, and injected instead of this
  anchor. Two additional fields will be added: `step` and `rev` (explained

>   above).

Reviewer 2: I don't see them explained above in this page.

- `<DVC_METRIC_TITLE>` (optional) - a title for the plot, that can be defined

>   with the `--title` option of the `dvc plot` subcommands.

Reviewer 2: ```suggestion
  with the `--title` option of the `dvc plots` subcommands.
```

- `<DVC_METRIC_X>` (optional) - field name of the data for the X axis. It can be
  defined with the `-x` option of the `dvc plot` subcommands. The auto-generated

>   `step` field (explained above) is the default.

Reviewer 2: Again, don't see it explained above.

- `<DVC_METRIC_Y>` (optional) - field name of the data for the Y axis. It can be
  defined with the `-y` option of the `dvc plot` subcommands. It defaults to the
  last header of the metrics file: the last column for CSV/TSV, or the last
  field for JSON/YAML.

- `<DVC_METRIC_X_LABEL>` (optional) - field name to display as the X axis label

- `<DVC_METRIC_Y_LABEL>` (optional) - field name to display as the Y axis label

## Options

- `-o <path>`, `--out <path>` - Directory to save templates to.

> ## Example: Modyfing simple template

Reviewer 2: ```suggestion
## Example: Modifying simple template
```

> We will modify the `simple` template to show bars instead of line. Lets take

Reviewer 2: ```suggestion
We will modify the `simple` template to show bars instead of lines. Lets take
```
